jax/_src/lax/parallel.py

```diff
@@ -1094,6 +1177,94 @@ def _raise_to_shaped_abstract_eval(x, *, axis_name, **params):
 batching.fancy_primitive_batchers[ppermute_p] = _ppermute_batcher
 batching.skippable_batchers[ppermute_p] = partial(_names_in_param, 'axis_name')
 
+
+def _psend_lowering(ctx, x, *, axis_name, perm):
+  if ("cuda" not in ctx.module_context.platforms):
+    raise NotImplementedError("psend is currently only implemented on GPUs")
+
+  full_perm, other_args = _pcollectives_lowering_common(
+      ctx, axis_name=axis_name, perm=perm, op_name="psend"
+  )
+  token = hlo.create_token()
+  send_op = hlo.SendOp(
+      [x],
+      token,
+      source_target_pairs=mlir.dense_int_elements(full_perm),
+      **other_args,
+  )
+  axis_ctx = ctx.module_context.axis_context
+  if isinstance(axis_ctx, ShardingContext):
```

Comment:
if not isinstance(axis_ctx, SPMDAxisContext):
  raise NotImplementedError(...)

Reply:
suggestion applied